### PR TITLE
Lasso and Elastic Net Optimization for Column-Major Tables

### PR DESCRIPTION
**Pull request by owerbat**
Some optimizations in Lasso, Elastic Net, and MSE kernels for SOA numeric tables

### CODE REVIEW

File: algorithms/kernel/objective_function/mse/mse_dense_default_batch_impl.i [hunk 1]
```diff
@@ -229,60 +170,36 @@ inline services::Status MSEKernel<algorithmFPType, method, cpu>::compute(Numeric
...
-                            });
-                            tlsData.reduce([&](algorithmFPType * localHes) {
+                            for (size_t i = 0; i < localBlockSize; i++)
+                            {
                                 PRAGMA_IVDEP
                                 PRAGMA_VECTOR_ALWAYS
                                 for (size_t j = 0; j < nTheta; j++)
                                 {
-                                    hessianDiagonalPtr[j] += localHes[j];
+                                    hessianDiagonalLocal[j] += x[i * dim + j] * x[i * dim + j]; /*USE DOTPRODUCT or parallel computation*/
```
Comment (agorshk) [inline algorithms/kernel/objective_function/mse/mse_dense_default_batch_impl.i:190]: Might be better to use BLAS dot product?

File: algorithms/kernel/objective_function/mse/mse_dense_default_batch_impl.i [hunk 2]
```diff
@@ -229,60 +170,36 @@ inline services::Status MSEKernel<algorithmFPType, method, cpu>::compute(Numeric
...
-                            });
-                        }
+                            }
+                        });
+                        tlsData.reduce([&](algorithmFPType * localHes) {
+                            PRAGMA_IVDEP
+                            PRAGMA_VECTOR_ALWAYS
+                            for (size_t j = 0; j < nTheta; j++)
+                            {
+                                hessianDiagonalPtr[j] += localHes[j];
```
Comment (agorshk) [inline algorithms/kernel/objective_function/mse/mse_dense_default_batch_impl.i:199]: xAXPY might be used

**Comment by agorshk** [comment]
I suggest to manually compare results with continuous and noncontinuous SOA data set, I'm not sure that in test system has such test. It's needed because the corresponding branches have been merged

File: algorithms/kernel/objective_function/mse/mse_dense_default_batch_impl.i [hunk 3]
```diff
@@ -62,64 +63,7 @@ inline services::Status MSEKernel<algorithmFPType, method, cpu>::compute(Numeric
         {
             if (xNT != dataNT)
```
Comment (ShvetsKS) [inline algorithms/kernel/objective_function/mse/mse_dense_default_batch_impl.i:64]: this condition doesn't make sense now, if code bellow is deleted.

File: algorithms/kernel/objective_function/mse/mse_dense_default_batch_impl.i [hunk 4]
```diff
@@ -62,64 +63,7 @@ inline services::Status MSEKernel<algorithmFPType, method, cpu>::compute(Numeric
...
-                    transposedData = false;
-                }
-                else if (soaDataPtr)
-                {
-                    const bool featuresEqual                            = soaDataPtr->getDictionary()->getFeaturesEqual();
-                    bool isContinuous                                   = false;
-                    data_management::features::IndexNumType currentType = features::internal::getIndexNumType<algorithmFPType>();
-                    if (currentType == (*(soaDataPtr->getDictionary()))[0].getIndexType())
-                    {
-                        isContinuous = true;
```
Comment (ShvetsKS) [inline algorithms/kernel/objective_function/mse/mse_dense_default_batch_impl.i:80]: Did you check performance on continuous columnar based input data (SOA table with continuous data)?

File: algorithms/kernel/objective_function/mse/mse_dense_default_batch_impl.i (hunk 1 above)
Comment (ShvetsKS) [reply algorithms/kernel/objective_function/mse/mse_dense_default_batch_impl.i:190]: yes, comment was waiting for the refactoring :)

**Comment by ShvetsKS** [comment]
Also please check performance degradation for case when nRows < nCols in case of homogen table.

File: algorithms/kernel/objective_function/mse/mse_dense_default_batch_impl.i [hunk 5]
```diff
@@ -229,60 +170,36 @@ inline services::Status MSEKernel<algorithmFPType, method, cpu>::compute(Numeric
...
-                                    PRAGMA_IVDEP
-                                    PRAGMA_VECTOR_ALWAYS
-                                    for (size_t j = 0; j < nTheta; j++)
-                                    {
-                                        hessianDiagonalLocal[j] += X[i * dim + j] * X[i * dim + j]; /*USE DOTPRODUCT or parallel computation*/
-                                    }
-                                }
-                            });
-                            tlsData.reduce([&](algorithmFPType * localHes) {
+                            for (size_t i = 0; i < localBlockSize; i++)
```
Comment (ShvetsKS) [inline algorithms/kernel/objective_function/mse/mse_dense_default_batch_impl.i:184]: ++i

File: algorithms/kernel/objective_function/mse/mse_dense_default_batch_impl.i [hunk 6]
```diff
@@ -229,60 +170,36 @@ inline services::Status MSEKernel<algorithmFPType, method, cpu>::compute(Numeric
...
+                            {
                                 PRAGMA_IVDEP
                                 PRAGMA_VECTOR_ALWAYS
                                 for (size_t j = 0; j < nTheta; j++)
```
Comment (ShvetsKS) [inline algorithms/kernel/objective_function/mse/mse_dense_default_batch_impl.i:188]: please correct here also ++j

File: algorithms/kernel/objective_function/mse/mse_dense_default_batch_impl.i [hunk 7]
```diff
@@ -229,60 +170,36 @@ inline services::Status MSEKernel<algorithmFPType, method, cpu>::compute(Numeric
...
-                        if (transposedData)
-                        {
-                            TlsMem<algorithmFPType, cpu, services::internal::ScalableCalloc<algorithmFPType, cpu> > tlsData(nTheta);
-                            daal::threader_for(nBlocks, nBlocks, [&](const size_t iBlock) {
-                                algorithmFPType * hessianDiagonalLocal = tlsData.local();
-                                const size_t startRow                  = iBlock * blockSize;
-                                const size_t finishRow                 = (iBlock + 1 == nBlocks ? nDataRows : (iBlock + 1) * blockSize);
+                        TlsMem<algorithmFPType, cpu, services::internal::ScalableCalloc<algorithmFPType, cpu> > tlsData(nTheta);
+                        daal::threader_for(nBlocks, nBlocks, [&](const size_t iBlock) {
+                            algorithmFPType * hessianDiagonalLocal = tlsData.local();
```
Comment (ShvetsKS) [inline algorithms/kernel/objective_function/mse/mse_dense_default_batch_impl.i:175]: algorithmFPType * const hessianDiagonalLocal

File: algorithms/kernel/lasso_regression/lasso_regression_train_dense_default_impl.i [hunk 8]
```diff
@@ -178,19 +174,35 @@ services::Status TrainBatchKernel<algorithmFPType, method, cpu>::compute(
...
         }
 
-        daal::threader_for(nBlocks, nBlocks, [&](const size_t iBlock) {
-            const size_t startRow  = iBlock * blockSize;
-            const size_t finishRow = (iBlock + 1 == nBlocks ? nRows : (iBlock + 1) * blockSize);
-            for (size_t i = startRow; i < finishRow; i++)
-            {
-                PRAGMA_IVDEP
-                PRAGMA_VECTOR_ALWAYS
+        SOANumericTable * soaPtr = dynamic_cast<SOANumericTable *>(xTrain.get());
```
Comment (ShvetsKS) [inline algorithms/kernel/lasso_regression/lasso_regression_train_dense_default_impl.i:177]: const

File: algorithms/kernel/lasso_regression/lasso_regression_train_dense_default_impl.i [hunk 9]
```diff
@@ -151,17 +146,18 @@ services::Status TrainBatchKernel<algorithmFPType, method, cpu>::compute(
...
+
+            for (size_t j = 0; j < nFeatures; j++)
             {
-                PRAGMA_IVDEP
-                PRAGMA_VECTOR_ALWAYS
-                for (size_t j = 0; j < nFeatures; j++)
-                {
-                    sum[j] += xPtr[i * nFeatures + j];
-                }
+                sum[j] += daal::internal::Blas<algorithmFPType, cpu>::xxdot(&numRowsInBlock, xPtr + j, &p, &ione, &zero);
```
Comment (ShvetsKS) [inline algorithms/kernel/lasso_regression/lasso_regression_train_dense_default_impl.i:160]: How much perf gain we obtain by usage `xxdot` here with parameters `&ione, &zero` instead off simple summation?

File: algorithms/kernel/lasso_regression/lasso_regression_train_dense_default_impl.i [hunk 10]
```diff
@@ -151,17 +146,18 @@ services::Status TrainBatchKernel<algorithmFPType, method, cpu>::compute(
 
         TlsMem<algorithmFPType, cpu, services::internal::ScalableCalloc<algorithmFPType, cpu> > tlsData(nFeatures);
         daal::threader_for(nBlocks, nBlocks, [&](const size_t iBlock) {
-            algorithmFPType * sum  = tlsData.local();
-            const size_t startRow  = iBlock * blockSize;
-            const size_t finishRow = (iBlock + 1 == nBlocks ? nRows : (iBlock + 1) * blockSize);
-            for (size_t i = startRow; i < finishRow; i++)
+            algorithmFPType * sum   = tlsData.local();
```
Comment (ShvetsKS) [inline algorithms/kernel/lasso_regression/lasso_regression_train_dense_default_impl.i:149]: `algorithmFPType * const sum`

File: algorithms/kernel/elastic_net/elastic_net_train_dense_default_impl.i [hunk 11]
```diff
@@ -178,19 +174,35 @@ services::Status TrainBatchKernel<algorithmFPType, method, cpu>::compute(
...
         }
 
-        daal::threader_for(nBlocks, nBlocks, [&](const size_t iBlock) {
-            const size_t startRow  = iBlock * blockSize;
-            const size_t finishRow = (iBlock + 1 == nBlocks ? nRows : (iBlock + 1) * blockSize);
-            for (size_t i = startRow; i < finishRow; i++)
-            {
-                PRAGMA_IVDEP
-                PRAGMA_VECTOR_ALWAYS
+        SOANumericTable * soaPtr = dynamic_cast<SOANumericTable *>(xTrain.get());
```
Comment (ShvetsKS) [inline algorithms/kernel/elastic_net/elastic_net_train_dense_default_impl.i:177]: const

File: algorithms/kernel/elastic_net/elastic_net_train_dense_default_impl.i [hunk 12]
```diff
@@ -151,17 +146,18 @@ services::Status TrainBatchKernel<algorithmFPType, method, cpu>::compute(
...
+            algorithmFPType * sum   = tlsData.local();
+            const size_t startRow   = iBlock * blockSize;
+            const size_t finishRow  = (iBlock + 1 == nBlocks ? nRows : (iBlock + 1) * blockSize);
+            DAAL_INT numRowsInBlock = finishRow - startRow;
+
+            daal::internal::ReadRows<algorithmFPType, cpu> xBD(xTrain.get(), startRow, numRowsInBlock);
+            DAAL_CHECK_BLOCK_STATUS_THR(xBD);
+            const algorithmFPType * xPtr = xBD.get();
+
+            for (size_t j = 0; j < nFeatures; j++)
```
Comment (ShvetsKS) [inline algorithms/kernel/elastic_net/elastic_net_train_dense_default_impl.i:158]: ++j

**Comment by ShvetsKS** [comment]
How a lot of failed tests in Precommit are related to changes in this PR?

**Pull request merged**